### PR TITLE
Fuzzy Runner

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Description
Finding and running commands is a pain. Arun had an idea that we could implement a fuzzy finding script

arun

Demo above

Implemented using seatgeeks thewuzz fuzzy finder and iterfzf

### Testing Done
Manual testing, encourage everyone to try it out

:warning: make sure to run setup software first before running `./tbots.py`

### Resolved Issues
N/A

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/tbots.py
Comment (nimazareian) [inline src/tbots.py]: Can we add `-s` (open to suggestions) for `--stop_ai_on_start` aswell?

File: src/tbots.py
Comment (nimazareian) [inline src/tbots.py]: guaranteed (edit: and appropriate) misspelled

File: src/tbots.py
Comment (nimazareian) [inline src/tbots.py]: by wuzz do you mean fuzz?

**Review by nimazareian — COMMENTED** [review]
I really like this idea, great job with implementing it so quick! 💯
Unfortunately Im not on Ubuntu to try it out right now, so I just got a few nits

**Review by nimazareian — COMMENTED** [review]
We could probably add an argument for building under debugger aswell

File: src/tbots.py
Comment (akhilveeraghanta) [reply src/tbots.py]: fixed

File: src/tbots.py
Comment (akhilveeraghanta) [reply src/tbots.py]: fixed

File: src/tbots.py
Comment (akhilveeraghanta) [reply src/tbots.py]: added

**Comment by akhilveeraghanta** [comment]
> We could probably add an argument for building under debugger aswell

added!

File: src/tbots.py
Comment (nimazareian) [inline src/tbots.py]: Is this meant to be `--stop_ai_on_start`?

File: src/tbots.py
Comment (nimazareian) [inline src/tbots.py:17]: nit: I'm not very familiar with `fuzz`, could you add a comment explaining what this represents?

File: src/tbots.py
Comment (akhilveeraghanta) [reply src/tbots.py]: good catch fixed!

File: src/tbots.py
Comment (akhilveeraghanta) [reply src/tbots.py:17]: added a comment

File: src/tbots.py
Comment (nimazareian) [inline src/tbots.py]: Last comment, nit: experimentally misspelled.

Comment (akhilveeraghanta) [reply src/tbots.py]: lol I can't spell for shit, fixed

Comment (nimazareian) [reply src/tbots.py]: I feel that aswell, you need a spell checker😅

**Review by nimazareian — DISMISSED** [review]
LGTM! Really excited about this. It should make the workflow much smoother

File: src/tbots.py
Comment (jonathanlew) [inline src/tbots.py]: it'd be nice if these weren't hard coded, i.e. what about passing in the network interface? What if we need to add new arguments in the future? If we had a `--args` argument to pass in a bunch of flags and then we had the C++ binary accept shortened flags, e.g. `-s`, then this script could be just pass the flags along to `test_args` or not depending on test vs run. Clearly, debug build, cache test results etc are good. So `tbots.py test cornerkick -d -p --args -s -v`

File: src/tbots.py
Comment (jonathanlew) [inline src/tbots.py:26]: why not always print out the command that it matched to for clarity? Also, I tried `./tbots.py test corner -s -v -p` and it matched to `bazel test //software/gui:geometry_conversion_test --cache_test_results=false --test_arg="--stop_ai_on_start" --test_arg="--enable_visualizer"
`

File: src/tbots.py
Comment (jonathanlew) [inline src/tbots.py:93]: why doesn't it run it?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
cool stuff

File: src/tbots.py
Comment (akhilveeraghanta) [reply src/tbots.py]: network interface arg is supported, the arguments that aren't listed here are passed through, see `parse_known_args` docs

so the only special cases are the two short hand ones, its harder to get the C++ binary to accept shorthand because the dynamic param schema doesn't have a spot for it. We could make one but that seems overkill for just these 2 special cases

File: src/tbots.py
Comment (akhilveeraghanta) [reply src/tbots.py:26]: hmm not sure what you mean? it does print the matched command rn for both selection and thefuzz

re: mismatch, yeah idk what the solution there is other than we might need to put a bit more effort in the query tbh, open to suggestions

File: src/tbots.py
Comment (akhilveeraghanta) [reply src/tbots.py:93]: it just gives you the command if you did `-p`

edit: to be a bit more clear, perhaps you wanna share the command with someone, you can just do -p to grab it and it won't run it
if you don't do -p, it'll run

File: src/tbots.py
Comment (nimazareian) [reply src/tbots.py:26]: Perhaps if this is happening often we would need to increase `THEFUZZ_MATCH_RATIO_THRESHOLD` to narrow down the search result accuracy

File: environment_setup/setup_software.sh
Comment (jonathanlew) [inline environment_setup/setup_software.sh:136]: also install python-Levenshtein for speed

File: src/tbots.py
Comment (jonathanlew) [reply src/tbots.py:93]: fair enough, see below for the change I'd like to see

File: src/tbots.py
Comment (jonathanlew) [inline src/tbots.py]: this would make it clear exactly what command is running:
```suggestion
    else:
        print(" ".join(command))
        os.system(" ".join(command))
```

File: src/tbots.py
Comment (jonathanlew) [reply src/tbots.py]: I see, a comment would help

File: src/tbots.py
Comment (jonathanlew) [reply src/tbots.py:26]: ok, this is fine

File: src/tbots.py
Comment (akhilveeraghanta) [reply src/tbots.py]: added

File: src/tbots.py
Comment (akhilveeraghanta) [reply src/tbots.py:93]: done

File: environment_setup/setup_software.sh
Comment (akhilveeraghanta) [reply environment_setup/setup_software.sh:136]: ah yes, I installed it locally but forgot to add it to the setup script
added now

**Review by nimazareian — DISMISSED** [review]
Looks good, pending CI approval

File: src/tbots.py
Comment (itsarune) [inline src/tbots.py:47]: I think we should have a way to select from the tests that exceed the confidence threshold

for instance: I ran ./tbots.py test penalty and I want to be able to select from

//software/ai/hl/stp/tactic/penalty_kick:penalty_kick_fsm_test
//software/ai/hl/stp/tactic/penalty_kick:penalty_kick_tactic_test
//software/ai/hl/stp/play:penalty_kick_enemy_play_test
//software/ai/hl/stp/play:penalty_kick_play_test

**Review by itsarune — COMMENTED** [review]
whipped up so fast, and looks really cool

File: src/tbots.py
Comment (akhilveeraghanta) [reply src/tbots.py:47]: interesting, so launch FZF for all tests matching confidence over some value?
or just do `./tbots.py test penaltyfsm` and `./tbots.py test penaltyenemy`

the way the wuzz matches it looks like they can give you n matches with decreasing match ratios, we can just do 1 and click the  "I'm feeling lucky" button equivalent and run it

or we can launch FZF everytime (I personally prefer having the "I'm feeling lucky" with more chars in the search query over having to select from fzf after a shorter search query)

thoughts?

Comment (itsarune) [reply src/tbots.py:47]: how about we can default to I'm feeling lucky and then maybe --list or -l flag to launch fzf

**Review by jonathanlew — COMMENTED** [review]
LGTM pending Arun's comment and CI

File: src/tbots.py
Comment (nimazareian) [reply src/tbots.py:47]: Another option would be to have "I'm feeling lucky" as the default, but if the top 3 results (can be different) have a confidence score with in 10 (also can be different) of eachother, we launch fzf? 
In the case of a common query like "test penalty" it could potentially make it clearer that there are multiple similar options, and in the case of other more specific queries with one main target, it would just run the top command.

Comment (akhilveeraghanta) [reply src/tbots.py:47]: added -i interactive flag

re: checking the confidence score, I tried to look at the values they don't seem to make too much sense like that, there are matches within 5 points that are not what we would be looking for

I've just changed it to trigger interactive mode if the confidence is below the threshold

**Review by jonathanlew — APPROVED** [review]
LGTM pending Arun's testing

**Review by nimazareian — APPROVED** [review]
LGTM!

File: src/tbots.py
Comment (itsarune) [reply src/tbots.py:47]: looks great

**Review by itsarune — APPROVED** [review]
awesome

**Pull request merged**
